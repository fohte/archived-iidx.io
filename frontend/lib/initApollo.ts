import 'isomorphic-fetch'

import { InMemoryCache, NormalizedCacheObject } from 'apollo-cache-inmemory'
import { ApolloClient } from 'apollo-client'
import { ApolloLink } from 'apollo-link'
import { setContext } from 'apollo-link-context'
import { onError } from 'apollo-link-error'
import { HttpLink } from 'apollo-link-http'

import { auth } from '../lib/firebaseApp'
import isBrowser from './isBrowser'

let apolloClient: ApolloClient<NormalizedCacheObject> | null = null

const handleError = onError(({ graphQLErrors, networkError }) => {
  if (graphQLErrors) {
    graphQLErrors.map(({ message, locations, path }) =>
      console.error(
        `[GraphQL error]: Message: ${message}, Location: ${JSON.stringify(
          locations,
        )}, Path: ${path}`,
      ),
    )
  }

  if (networkError) {
    console.error(`[Network error]: ${networkError}`)
  }
})

const withAuthorization = setContext(async (_, { headers }) => {
  if (auth.currentUser) {
    const idToken = await auth.currentUser.getIdToken()

    if (idToken) {
      return {
        headers: { ...headers, authorization: `Bearer ${idToken}` },
      }
    }
  }

  return { headers }
})

const create = (initialState?: NormalizedCacheObject) => {
  const cache = new InMemoryCache().restore(initialState || {})

  const httpLink = new HttpLink({
    uri: process.env.API_URL,
    credentials: 'same-origin',
  })

  return new ApolloClient({
    connectToDevTools: isBrowser,
    ssrMode: !isBrowser,
    link: ApolloLink.from([handleError, withAuthorization, httpLink]),
    cache,
  })
}

const initApollo = (initialState?: NormalizedCacheObject) => {
  // Make sure to create a new client for every server-side request so that data
  // isn't shared between connections (which would be bad)
  if (!isBrowser) {
    return create(initialState)
  }

  // Reuse client on the client-side
  if (!apolloClient) {
    apolloClient = create(initialState)
  }

  return apolloClient
}

export default initApollo
