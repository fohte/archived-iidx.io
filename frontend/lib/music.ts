export const makeTitle = (music: { title: string; subTitle: string }) =>
  `${music.title} ${music.subTitle}`
