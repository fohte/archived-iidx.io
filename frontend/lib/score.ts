import * as _ from 'lodash'

export enum Grade {
  Max = 'MAX',
  A = 'A',
  AA = 'AA',
  AAA = 'AAA',
  B = 'B',
  C = 'C',
  D = 'D',
  E = 'E',
  F = 'F',
}

export const calcScoreRate = (
  score: number,
  numNotes: number,
  precision: number = 2,
) => _.round((score / (numNotes * 2)) * 100, precision)

const getGradeBorders = (numNotes: number): { [key in Grade]: number } => {
  const maxScore = numNotes * 2
  const calcGradeBorder = (coefficients: number): number =>
    Math.ceil(maxScore * coefficients)

  return {
    [Grade.F]: 0,
    [Grade.E]: calcGradeBorder(2 / 9),
    [Grade.D]: calcGradeBorder(3 / 9),
    [Grade.C]: calcGradeBorder(4 / 9),
    [Grade.B]: calcGradeBorder(5 / 9),
    [Grade.A]: calcGradeBorder(6 / 9),
    [Grade.AA]: calcGradeBorder(7 / 9),
    [Grade.AAA]: calcGradeBorder(8 / 9),
    [Grade.Max]: maxScore,
  }
}

export const searchGrade = (
  score: number,
  numNotes: number,
): [Grade | null, number] => {
  if (numNotes * 2 === score) {
    return [Grade.Max, 0]
  }

  const gradeBorders = getGradeBorders(numNotes)

  const diffs = _.mapValues(gradeBorders, border => score - border)
  const diffArray = _.toPairs(diffs) as Array<[Grade, number]>

  const nearest = _.minBy<[Grade, number]>(
    _.filter(diffArray, ([, diff]) => diff >= 0),
    ([, diff]) => Math.abs(diff),
  )

  return nearest || [null, 0]
}

export const searchNextGrade = (
  score: number,
  numNotes: number,
): [Grade | null, number] => {
  if (numNotes * 2 === score) {
    return [Grade.Max, 0]
  }

  const gradeBorders = getGradeBorders(numNotes)

  const diffs = _.mapValues(gradeBorders, border => score - border)
  const diffArray = _.toPairs(diffs) as Array<[Grade, number]>

  const nearest = _.minBy<[Grade, number]>(
    _.filter(diffArray, ([grade, diff]) => diff < 0 && grade !== Grade.F),
    ([, diff]) => Math.abs(diff),
  )

  return nearest || [null, 0]
}
