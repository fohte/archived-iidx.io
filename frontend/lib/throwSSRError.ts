import { ServerResponse } from 'http'

export default (res: ServerResponse | undefined, statusCode: number) => {
  if (res) {
    res.statusCode = statusCode
  }
}
