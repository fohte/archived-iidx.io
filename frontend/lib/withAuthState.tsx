import { setDisplayName, wrapDisplayName } from 'recompose'
import { Diff } from 'utility-types'

import AuthContext, { AuthContextShape } from '@app/contexts/AuthContext'

export interface ExternalProps {}
export type InjectedProps = AuthContextShape

const withAuthState = () => <OriginalProps extends {}>(
  WrappedComponent: React.ComponentType<OriginalProps & InjectedProps>,
) => {
  type EnhancedProps = Diff<OriginalProps, InjectedProps> & ExternalProps

  const WithAuthState: React.SFC<EnhancedProps> = props => (
    <AuthContext.Consumer>
      {authContexts => <WrappedComponent {...props} {...authContexts} />}
    </AuthContext.Consumer>
  )

  const newDisplayName = wrapDisplayName(WrappedComponent, 'withAuthState')

  return setDisplayName<EnhancedProps>(newDisplayName)(WithAuthState)
}

export default withAuthState
