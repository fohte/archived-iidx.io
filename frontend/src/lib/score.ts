import * as _ from 'lodash'

export enum Grade {
  Max = 'MAX',
  A = 'A',
  AA = 'AA',
  AAA = 'AAA',
  B = 'B',
  C = 'C',
  D = 'D',
  E = 'E',
  F = 'F',
}

export interface GradeDiff {
  grade: Grade | null
  diff: number
}

export const calcScoreRate = (
  score: number,
  numNotes: number,
  precision: number = 2,
) => _.round((score / (numNotes * 2)) * 100, precision)

export const calcGradeBorder = (coefficients: number, numNotes: number) =>
  Math.ceil(numNotes * 2 * coefficients)

export const getGradeBorders = (
  numNotes: number,
): { [key in Grade]: number } => ({
  [Grade.F]: 0,
  [Grade.E]: calcGradeBorder(numNotes, 2 / 9),
  [Grade.D]: calcGradeBorder(numNotes, 3 / 9),
  [Grade.C]: calcGradeBorder(numNotes, 4 / 9),
  [Grade.B]: calcGradeBorder(numNotes, 5 / 9),
  [Grade.A]: calcGradeBorder(numNotes, 6 / 9),
  [Grade.AA]: calcGradeBorder(numNotes, 7 / 9),
  [Grade.AAA]: calcGradeBorder(numNotes, 8 / 9),
  [Grade.Max]: numNotes * 2,
})

export const defaultGradeDiff: GradeDiff = {
  grade: null,
  diff: 0,
}

export const searchGrade = (score: number, numNotes: number): GradeDiff => {
  if (numNotes * 2 === score) {
    return { grade: Grade.Max, diff: 0 }
  }

  const gradeBorders = getGradeBorders(numNotes)

  const diffs = _.mapValues(gradeBorders, border => score - border)
  const diffArray = _.toPairs(diffs) as Array<[Grade, number]>

  const nearest = _.minBy<[Grade, number]>(
    _.filter(diffArray, ([, diff]) => diff >= 0),
    ([, diff]) => Math.abs(diff),
  )

  if (!nearest) {
    return defaultGradeDiff
  }

  return { grade: nearest[0], diff: nearest[1] }
}

export const searchNextGrade = (score: number, numNotes: number): GradeDiff => {
  if (numNotes * 2 === score) {
    return { grade: Grade.Max, diff: 0 }
  }

  const gradeBorders = getGradeBorders(numNotes)

  const diffs = _.mapValues(gradeBorders, border => score - border)
  const diffArray = _.toPairs(diffs) as Array<[Grade, number]>

  const nearest = _.minBy<[Grade, number]>(
    _.filter(diffArray, ([grade, diff]) => diff < 0 && grade !== Grade.F),
    ([, diff]) => Math.abs(diff),
  )

  if (!nearest) {
    return defaultGradeDiff
  }

  return { grade: nearest[0], diff: nearest[1] }
}

export const isMax = ({ grade, diff }: GradeDiff): boolean =>
  grade === Grade.Max && diff === 0
