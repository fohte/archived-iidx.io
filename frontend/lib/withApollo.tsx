import { NormalizedCacheObject } from 'apollo-cache-inmemory'
import { ApolloClient } from 'apollo-client'
import Head from 'next/head'
import React from 'react'
import { ApolloProvider, getDataFromTree } from 'react-apollo'
import { setDisplayName, wrapDisplayName } from 'recompose'

import initApollo from './initApollo'
import isBrowser from './isBrowser'

interface Props {
  serverState: {
    apollo: {
      data: NormalizedCacheObject
    }
  }
}

export default ComposedComponent => {
  class WithData extends React.Component<Props> {
    public static async getInitialProps(ctx) {
      // Initial serverState with apollo (empty)
      let serverState: Props['serverState'] = {
        apollo: {
          data: {},
        },
      }

      // Evaluate the composed component's getInitialProps()
      let composedInitialProps = {}
      if (ComposedComponent.getInitialProps) {
        composedInitialProps = await ComposedComponent.getInitialProps(ctx)
      }

      // Run all GraphQL queries in the component tree
      // and extract the resulting data
      if (!isBrowser) {
        const apollo = initApollo()

        try {
          // Run all GraphQL queries
          await getDataFromTree(
            <ApolloProvider client={apollo}>
              <ComposedComponent {...composedInitialProps} />
            </ApolloProvider>,
            {
              router: {
                asPath: ctx.asPath,
                pathname: ctx.pathname,
                query: ctx.query,
              },
            },
          )
        } catch (error) {
          // Prevent Apollo Client GraphQL errors from crashing SSR.
          // Handle them in components via the data.error prop:
          // https://www.apollographql.com/docs/react/api/react-apollo.html#graphql-query-data-error
        }
        // getDataFromTree does not call componentWillUnmount
        // head side effect therefore need to be cleared manually
        Head.rewind()

        // Extract query data from the Apollo store
        serverState = {
          apollo: {
            data: apollo.cache.extract(),
          },
        }
      }

      return {
        serverState,
        ...composedInitialProps,
      }
    }

    public apollo: ApolloClient<NormalizedCacheObject>

    constructor(props) {
      super(props)
      this.apollo = initApollo(this.props.serverState.apollo.data)
    }

    public render() {
      return (
        <ApolloProvider client={this.apollo}>
          <ComposedComponent {...this.props} />
        </ApolloProvider>
      )
    }
  }

  return setDisplayName(wrapDisplayName(WithData, 'WithData'))(WithData)
}
