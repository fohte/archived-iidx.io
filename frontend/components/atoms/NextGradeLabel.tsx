import * as React from 'react'
import { Label } from 'semantic-ui-react'

import { formatNumber } from '@app/components/atoms/GradeLabel'
import { Grade } from '@app/lib/score'

export interface Props {
  grade: Grade
  diff: number
}

const NextGradeLabel: React.SFC<Props> = ({ grade, diff }) => (
  <Label horizontal>
    Next
    <Label.Detail>
      {grade} {formatNumber(diff)}
    </Label.Detail>
  </Label>
)

export default NextGradeLabel
