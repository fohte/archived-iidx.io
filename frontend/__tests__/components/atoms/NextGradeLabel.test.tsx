import * as _ from 'lodash'
import * as React from 'react'
import * as renderer from 'react-test-renderer'

import NextGradeLabel from '@app/components/atoms/NextGradeLabel'
import { Grade } from '@app/lib/score'

describe('NextGradeLabel', () => {
  _.forEach(Grade, grade => {
    it(`renders correctly when clear lamp is ${grade}`, () => {
      const tree = renderer
        .create(<NextGradeLabel grade={grade} diff={0} />)
        .toJSON()
      expect(tree).toMatchSnapshot()
    })
  })

  it('renders correctly when diff is greater than 0', () => {
    const tree = renderer
      .create(<NextGradeLabel grade={Grade.AAA} diff={1} />)
      .toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('renders correctly when diff is 0', () => {
    const tree = renderer
      .create(<NextGradeLabel grade={Grade.AAA} diff={0} />)
      .toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('renders correctly when diff is less than 0', () => {
    const tree = renderer
      .create(<NextGradeLabel grade={Grade.AAA} diff={-1} />)
      .toJSON()
    expect(tree).toMatchSnapshot()
  })
})
