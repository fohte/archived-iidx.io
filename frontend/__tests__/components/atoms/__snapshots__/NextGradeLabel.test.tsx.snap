// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NextGradeLabel renders correctly when clear lamp is A 1`] = `
<div
  className="ui horizontal label"
  onClick={[Function]}
>
  A
  <div
    className="detail"
  >
    0
  </div>
</div>
`;

exports[`NextGradeLabel renders correctly when clear lamp is AA 1`] = `
<div
  className="ui horizontal label"
  onClick={[Function]}
>
  AA
  <div
    className="detail"
  >
    0
  </div>
</div>
`;

exports[`NextGradeLabel renders correctly when clear lamp is AAA 1`] = `
<div
  className="ui horizontal label"
  onClick={[Function]}
>
  AAA
  <div
    className="detail"
  >
    0
  </div>
</div>
`;

exports[`NextGradeLabel renders correctly when clear lamp is B 1`] = `
<div
  className="ui horizontal label"
  onClick={[Function]}
>
  B
  <div
    className="detail"
  >
    0
  </div>
</div>
`;

exports[`NextGradeLabel renders correctly when clear lamp is C 1`] = `
<div
  className="ui horizontal label"
  onClick={[Function]}
>
  C
  <div
    className="detail"
  >
    0
  </div>
</div>
`;

exports[`NextGradeLabel renders correctly when clear lamp is D 1`] = `
<div
  className="ui horizontal label"
  onClick={[Function]}
>
  D
  <div
    className="detail"
  >
    0
  </div>
</div>
`;

exports[`NextGradeLabel renders correctly when clear lamp is E 1`] = `
<div
  className="ui horizontal label"
  onClick={[Function]}
>
  E
  <div
    className="detail"
  >
    0
  </div>
</div>
`;

exports[`NextGradeLabel renders correctly when clear lamp is F 1`] = `
<div
  className="ui horizontal label"
  onClick={[Function]}
>
  F
  <div
    className="detail"
  >
    0
  </div>
</div>
`;

exports[`NextGradeLabel renders correctly when clear lamp is MAX 1`] = `
<div
  className="ui horizontal label"
  onClick={[Function]}
>
  MAX
  <div
    className="detail"
  >
    0
  </div>
</div>
`;

exports[`NextGradeLabel renders correctly when diff is 0 1`] = `
<div
  className="ui horizontal label"
  onClick={[Function]}
>
  AAA
  <div
    className="detail"
  >
    0
  </div>
</div>
`;

exports[`NextGradeLabel renders correctly when diff is greater than 0 1`] = `
<div
  className="ui horizontal label"
  onClick={[Function]}
>
  AAA
  <div
    className="detail"
  >
    +1
  </div>
</div>
`;

exports[`NextGradeLabel renders correctly when diff is less than 0 1`] = `
<div
  className="ui horizontal label"
  onClick={[Function]}
>
  AAA
  <div
    className="detail"
  >
    -1
  </div>
</div>
`;
